src/storage-operators/src/oneshot_source.rs

```diff
@@ -0,0 +1,880 @@
+
+                // Process each stream of work, one at a time.
+                for maybe_request in maybe_requests {
+                    let request = maybe_request?;
+
+                    let mut work_stream = format.fetch_work(&source, request);
+                    while let Some(result) = work_stream.next().await {
+                        let record_chunk = result.context("fetch worker")?;
+                        record_chunks.push(record_chunk);
+                    }
+                }
+
+                Ok::<_, StorageErrorX>(record_chunks)
+            }
+            .await
+            .context("fetch work");
+
+            match result {
+                Ok(record_chunks) => record_chunks
+                    .into_iter()
+                    .for_each(|chunk| record_handle.give(&capability, Ok(chunk))),
+                Err(err) => record_handle.give(&capability, Err(err)),
+            }
+        }
+    });
+
+    (record_stream, shutdown.press_on_drop())
+}
+
+/// Render an operator that given a stream of [`OneshotFormat::RecordChunk`]s will decode these
+/// chunks into a stream of [`Row`]s.
+pub fn render_decode_chunk<G, F>(
+    scope: G,
+    format: F,
+    record_chunks: &TimelyStream<G, Result<F::RecordChunk, StorageErrorX>>,
+) -> (
+    TimelyStream<G, Result<Row, StorageErrorX>>,
+    PressOnDropButton,
+)
+where
+    G: Scope,
+    F: OneshotFormat + 'static,
+{
+    let mut builder = AsyncOperatorBuilder::new("CopyFrom-decode_chunk".to_string(), scope.clone());
+
+    let (row_handle, row_stream) = builder.new_output::<CapacityContainerBuilder<_>>();
+    let mut record_chunk_handle = builder.new_input_for(record_chunks, Distribute, &row_handle);
+
+    let shutdown = builder.build(move |caps| async move {
+        let [_row_cap] = caps.try_into().unwrap();
+
+        while let Some(event) = record_chunk_handle.next().await {
+            let (capability, maybe_chunks) = match event {
+                AsyncEvent::Data(cap, data) => (cap, data),
+                AsyncEvent::Progress(_) => continue,
+            };
+
+            let result = async {
+                let mut rows = Vec::new();
+                for maybe_chunk in maybe_chunks {
+                    let chunk = maybe_chunk?;
+                    format.decode_chunk(chunk, &mut rows)?;
+                }
+                Ok::<_, StorageErrorX>(rows)
+            }
+            .await
+            .context("decode chunk");
+
+            match result {
+                Ok(rows) => rows
+                    .into_iter()
+                    .for_each(|row| row_handle.give(&capability, Ok(row))),
+                Err(err) => row_handle.give(&capability, Err(err)),
+            }
+        }
+    });
+
+    (row_stream, shutdown.press_on_drop())
+}
+
+/// Render an operator that given a stream of [`Row`]s will stage them in Persist and return a
+/// stream of [`ProtoBatch`]es that can later be linked into a shard.
+pub fn render_stage_batches_operator<G>(
```

Comment:
I have a similar resuability question here. The persist_sink implementation we have in storage already has a stage that writes batches to persist and then collects the batch descriptions onto one worker to perform the CaA call. I was expecting that we'd reuse those operator and only replace the last append operator with one that sends the descriptions back to the controller.

Reply:
I read through the `write_batches` operator in the `persist_sink` and opted not to use it because similar to the `SourceRender` trait it seems to be setup to do much more than we require. I'm more optimistic we could refactor this code to use `write_batches`, but given that `render_stage_batches_operator` from this PR is essentially just a loop that appends to a Persist Batch, the amount of code duplication isn't that large, so a new operator seemed okay to me.